Updating C++ style guidelines for optionals, early returns, and assertions (#15)
Expensify/Style-Guides: cpp.md

auto name = result[0]["name"];
auto value = result[0]["value"];
```

## Unpacking optionals

When unpacking an optional value, you should use the `.value()` method or `.value_or()` rather than directly dereferencing with `*`

```cpp
const optional<string> myOptional = optional{"Hello world!"};

// Good
const string good = myOptional.value();

// Bad
const string bad = *myOptional;
```

When making assertions about whether or not an optional has a value, use `.has_value()` rather than relying on its implicit truthiness.

```cpp
const optional<string> myOptional = optional{"Hello world!"};

// Bad - implicit conversion of optional to boolean
if (myOptional) {
    doSomething();
}

// Good - clear unpacking of optional
if (myOptional.has_value()) {
    doSomething();
}

// Bad - implicit conversion of optional to boolean
ASSERT_FALSE(myOptional);

// Good - clear unpacking of optional
ASSERT_FALSE(myOptional.has_value());
```

## Prefer early returns

> Just like in E/App, we should prefer early returns when it's feasible to do so:

Reviewer: I didn't know we had this in App as well, but it's in our general guidelines https://github.com/Expensify/Style-Guide/blob/main/guidelines.md#best-practices-to-follow. I didn't think we need it here, but I guess there's no harm in having it in all places

```cpp

> void myGoodFunc(const string& myStr)
> {
>     // good, early return
>     if (myStr.empty()) {
>         return;
>     }
>     doSomething(myStr);
> }
> 
> void myBadFunc(const string& myStr)
> {
>     // bad, effects nested in conditional
>     if (!myStr.empty()) {
>         doSomething(myStr);
>     }
> }

Reviewer: ```suggestion
void myGoodFunction(const string& name)
{
    // Good, early return
    if (name.empty()) {
        return;
    }
    doSomething(name);
}

void myBadFunction(const string& name)
{
    // Bad, nested functionality
    if (!myStr.empty()) {
        doSomething(name);
    }
}
```

```

## Using assertions in tests

> You should prefer using `ASSERT_` over `EXPECT_` in tests, unless you have a specific reason to do otherwise. The reason for this is because `ASSERT_` will fail the test with the first failed assertion, while `EXPECT_` will fail the test and keep going, making the root cause of the failure less clear.

Reviewer: I think the best practice should be a bit more nuanced. I don't know where the convo with Flo is, but I think it's better to prefer EXPECT **unless** the subsequent assertions depend on the current one.

I think when using EXPECT the failures are pretty clear, as on dev they show the line where it failed and all.

I might be wrong about some thing here, but how's this?

```suggestion
You should prefer using `EXPECT_` over `ASSERT_` in tests, given that when `EXPECT_` assertion fails the execution of the test doesn't stop. This is useful when working on tests because you don't have to get one working in order to get the next one to run. Moreover, when assertions fail using `EXPECT_` they output shows the assertion that failed with its line number.

You should use `ASSERT_` for assertions that need to pass before the next assertions in the test are run. Said differently, if one assertion depends on another one succeeding, you should use `ASSERT_` on the first one because there's no point in continuing to run the test if the `ASSERT_` fails
```

```cpp

> // GOOD
> ASSERT_TRUE(true);
> 
> // Bad
> EXPECT_TRUE(true);

Reviewer: ```suggestion
// Good
ASSERT_TRUE(expression);

// Bad
EXPECT_TRUE(expression);
```
